Fixes an Index Checker bug and a related viewpoint adaptation bug

kelloggm opened the pull request:
This fixes kelloggm#139. There are two parts to the fix, but I'm PR'ing them together because either one is insufficient to fix the whole bug:

First, the SameLen Checker both actively assigns the wrong type to the right side of an assignment like `int @SameLen("otherArray") [] x = new int[otherArray.length]` (it before assigned that `@SameLen("new int[]")` via the common assignment checks, which is obviously incorrect), and fails to assign the proper type to the right hand side (i.e. the `new int...` part). These are solved by the changes to SameLenVisitor and SameLenAnnotatedTypeFactory, respectively.

Applying that fix causes a viewpoint adaptation bug to appear in the regression test case [`SLSubtyping.java` on line 22](https://github.com/typetools/checker-framework/blob/master/checker/tests/index/SLSubtyping.java). The change to `GenericAnnotatedTypeFactory` fixes that bug. The left side of that assignment has a user-written annotation (`@SameLen("f")`) and the right side is `f`, but `f` is a field, so the test was failing and claiming that `@SameLen("f")` and `@SameLen("(this).f")` were not the same.

mernst commented on checker/src/org/checkerframework/checker/index/samelen/SameLenAnnotatedTypeFactory.java:
```diff
@@ -234,6 +242,44 @@ public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+                new ImplicitsTreeAnnotator(this));
+    }
+
+    protected class SameLenTreeAnnotator extends TreeAnnotator {
```
Could you add documentation to indicate what this is intended to do?  The documentation might go here or elsewhere.

kelloggm replied:
I've added a comment that describes why a tree annotator is needed.

wmdietlGC commented on checker/tests/index/SameLenNewArrayWithSameLength.java:
```diff
@@ -0,0 +1,7 @@
+
+class SameLenNewArrayWithSameLength {
+    public void m1(int[] a) {
+        int @SameLen("a") [] b = new int[a.length]; // index checker warning
```
The `index checker warning` is presumably before your fix. If it is expected, it should be a proper expected output. If not, reformulate to make clear that there used to be a warning.

wmdietlGC commented on framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java:
```diff
@@ -1195,6 +1195,10 @@ public AnnotatedTypeMirror getAnnotatedTypeLhs(Tree lhsTree) {
             case MEMBER_SELECT:
             case ARRAY_ACCESS:
                 res = getAnnotatedType(lhsTree);
+                if (dependentTypesHelper != null && lhsTree.getKind() == Kind.VARIABLE) {
```
Is this really the right place to perform this standardization? Shouldn't it be part of `getAnnotatedType`?

smillst replied:
The code is wrong.  The problem is not calling super in createTreeAnnotator method.

smillst replied:
Also, in general, if you think you found a framework bug, it's good to confirm it using more than one checker.

smillst commented on checker/src/org/checkerframework/checker/index/samelen/SameLenAnnotatedTypeFactory.java:
```diff
@@ -234,6 +242,48 @@ public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                new SameLenTreeAnnotator(this),
+                new PropagationTreeAnnotator(this),
```
Call super!  You are missing the DependentTypesTreeAnnotator!

smillst commented on checker/tests/index/SameLenNewArrayWithSameLength.java:
```diff
@@ -0,0 +1,7 @@
+import org.checkerframework.checker.index.qual.*;
+
+class SameLenNewArrayWithSameLength {
+    public void m1(int[] a) {
```
Please add test cases where `a` is the same length of other arrays (besides `b`).

smillst commented on checker/src/org/checkerframework/checker/index/samelen/SameLenVisitor.java:
```diff
@@ -38,6 +38,7 @@ protected void commonAssignmentCheck(
             /*@CompilerMessageKey*/ String errorKey) {
         if (valueType.getKind() == TypeKind.ARRAY
                 && TreeUtils.isExpressionTree(valueTree)
+                && !(valueTree.getKind() == Tree.Kind.NEW_ARRAY)
```
On line 52 (below), call `isReceiverToStringParsable(rec)` instead of `!(rec instanceof Unknown)`.

kelloggm commented:
I've addressed all review comments and reverted the incorrect change to `GenericAnnotatedTypeFactory`. @smillst can you take a look at the code again? Thanks.
